### PR TITLE
429 catch added with other error codes in serp_goog

### PR DESCRIPTION
**Pull request by maximrus23**
Hi, 

I have used your library in the research purposes and it is great since there are no good docs for the Search API. However, it broke after QPS was exceeded from my IP and I got 429. In the change for safety, I have set 0.1 sec wait for each request and 0.5 if there is an error. For most of the time, your code as it is won't fail, but it is not robust in this small bit. 
Googleapiclient dependency is probably needed.

EDIT - search API has 100 requests per minute meaning 1.7 QPS. Adjust sleep time to that limit maybe in except statement. Otherwise, also possible to time the execution after request and add difference as waiting time before each request to speed up the script and catch exceptions

### CODE REVIEW

**Comment by maximrus23** [comment]
One small update - search API has 100 requests per minute meaning 1.7 QPS. Adjust sleep time to that limit maybe in except statement. Otherwise, also possible to time the execution after request and add difference as waiting time before each request to speed up the script and catch exceptions

**Comment by eliasdabbas** [comment]
Thanks @maximrus23 
Appreciate your suggestions and code. 

I went through your code, and here are some thoughts. 


`userRateLimitExceeded` is definitely something we can fix with your suggestion (see below).

I'm not sure why we should repeat the request after receiving an error? 
For example for `quotaExceeded` error needs to be changed manually through the API dashboard, so repeating the request would still produce an error. 
`internalServerError` and `backendError` are not clear, so maybe the 0.5 seconds won't be enough. 


The idea behind interrupting the process is that I don't want to waste the user's time/budget. If there is an error it should be handled/fixed/reported immediately. Better to be slightly disappointed now, instead of getting more disappointed later after making hundreds of requests for example. 

Errors are returned as they are from Google, so the user can fix or properly report (maybe there are other errors that we haven't thought of): 

Screen Shot 2020-06-24 at 1 50 05 PM



I like your edit. Instead of adding a time delay to each request, which would make it really long, we can use an approach like this:

```
if num_requests >= 98: 
    if elapsed time  >= 58:
        time.sleep(the_difference)
        reset_time
        num_requests = 0
```
Of course we would have to set up the start time, and a counter for the current number of requests. 

To summarize, I think it's a great idea to handle the number of requests against time, and `time.sleep` only when needed, as much as needed.  
For errors, I think we should just interrupt the process and return the error as it was received from Google.  

Let me know what you think, and if I missed anything. 
Thanks again :)

**Comment by maximrus23** [comment]
Hi @eliasdabbas 
Yeah, everything is relevant and exception handling should be added if only ```userRateLimitExceeded``` caught.  Adding your suggestions with elapsed time and requests would work great and won't even throw an exception.
Overall, completely agree with your thoughts!

**Comment by eliasdabbas** [comment]
@maximrus23 
Great!  
Would you like to make the modifications, or shall I handle this?

**Comment by maximrus23** [comment]
Hi @eliasdabbas . I came up with this instead of yours logic: 
```python
n, resp_num, timer = [0 for i in range(3)]

responses = []
for param in params_list:
    param_log = ', '.join([k + '=' + str(v) for k, v in param.items()])
    logging.info(msg='Requesting: ' + param_log)
# try block sends requests up 100 requests and 60 seconds
   resp_num += 1
   if ((resp_num <= 100) and (timer >= 60)):
       timer = 0
       resp_num = 0
                
   elif ((resp_num == 100) and (timer < 60)):
       time.sleep(60 - timer)
       timer = 0
       resp_num = 0
                
        tic = time.time()
        resp = requests.get(base_url, params=param)
        responses.append(resp)
        toc = time.time()
        # timer is set here since we also time append operation
        timer += (toc - tic)
```
The logic is: ```if``` resets timer in case user requested less than 100 in a minute. ```Elif``` on the contrary checks if the limit is exhausted. If it is, then it waits the difference and proceeds to requests with reset timer and num_requests. 
In the case below, I am not sure how to update the timer, for example,  num_requests can be < 50 and elapsed_time > 70. 
 ``` 
if num_requests >= 98: 
    if elapsed time  >= 58:
          ....
```
Feel free to modify and commit. Peace:)

**Comment by eliasdabbas** [comment]
Thanks again. 

* `timer` was not a good name for a variable, and `tic` / `toc` seems clearer as to its intention. 
* While comparing the time I made a mistake it should check if the elapsed time is less than 58 and not the other way around.
* Comparing to 100 requests and 60 seconds is "more correct" but I prefer to compare to 98 and 58 to be on the safe side:
    1. The actual counter and timer are not running on the user's computer, and there is a delay in requests and responses (we can't know exactly when it started)
    2. This is a very rare case, but if/when it happens, it's ok to delay one or two extra seconds for more safety

The only thing we need to do is check every 98 requests, whether or not the time elapsed since the "zeroth" request is less than a minute, which means we are about to hit the quota. In this case we `time.sleep` 60 - `elapsed`, and then reset `tic` and reset `num_requests` and the cycle starts over, as many times as needed. 


`if ((resp_num <= 100) and (timer >= 60)):`

Checking if it is <= 100 and > 60 is not something we need to worry about, because we would be well within our limits. 

If the below logic is correct, then we don't even need to `try` / `except` because we are preventing a certain error from happening in the first place. Other errors will raise, as we previously agreed. 

This is very similar to your logic just without the first condition, and without `timer`:

```python
num_requests = 0
tic = time.time()

responses = []

for param in params_list:
    resp = requests.get(base_url, params=param)
    responses.append(resp)
    num_requests += 1
    if num_requests == 98:
        toc = time.time()
        elapsed = toc - tic
        if elapsed < 58:
            time.sleep(60  - elspsed)
        num_requests = 0
        tic = time.time()

```

Does it make sense, did I miss anything?

**Comment by maximrus23** [comment]
@eliasdabbas Yeah, all pretty logical:)

**Comment by eliasdabbas** [comment]
@maximrus23 
Great! 

Ok, so I just want to make sure this pull request is in your name. Would you mind copying the last code that we agreed on and submitting a new PR? 

I'm also thinking of another thing, in addition to what we agreed. 
If an error occurs, break the loop, append an error message to the final df, and return the current df (with whatever data was retreived so far). This way, the user gets the data that didn't produce errors, and they get the error message as well as the final row in the result df.
